R/as_model_out_df.R

```diff
@@ -0,0 +1,211 @@
+  old_names <- names(tbl)
+
+  if (!is.null(model_id_col)) {
+    model_id_col <- validate_col_input(model_id_col, call = call)
+    tbl <- rename_col(tbl, model_id_col, old_names, call)
+  }
+  if (!is.null(output_type_col)) {
+    output_type_col <- validate_col_input(output_type_col, call = call)
+    tbl <- rename_col(tbl, output_type_col, old_names, call)
+  }
+  if (!is.null(output_type_id_col)) {
+    output_type_id_col <- validate_col_input(output_type_id_col, call = call)
+    tbl <- rename_col(tbl, output_type_id_col, old_names, call)
+  }
+  if (!is.null(value_col)) {
+    value_col <- validate_col_input(value_col, call = call)
+    tbl <- rename_col(tbl, value_col, old_names, call)
+  }
+  return(tbl)
+}
+
+validate_col_input <- function(x, call = rlang::caller_env()) {
+  if (!is.character(x)) {
+    cli::cli_abort("{.arg {rlang::caller_arg(x)}} must be a {.cls character} string
+                       instead of {.cls {typeof(x)}}",
+      call = call
+    )
+  }
+  if (length(x) > 1L) {
+    cli::cli_warn("{.arg {rlang::caller_arg(x)}} must be character vector of
+                       length {.val {1L}} not length {.val {length(x)}}.
+                      Only element 1 ({.val {x[1]}}) used, the rest ignored.",
+      call = call
+    )
+    x <- x[1]
+  }
+  return(x)
+}
+
+rename_col <- function(x, col_name, old_names, call) {
+  arg_name <- rlang::caller_arg(col_name)
+  if (!col_name %in% names(x)) {
+    cli::cli_abort(c(
+      "x" = "{.arg {arg_name}} value {.val {col_name}}
+                       not a valid column name in {.arg x}",
+      "!" = "Must be one of {.val {old_names}}"
+    ),
+    call = call
+    )
+  } else {
+    new_col_name <- gsub("_col", "", arg_name)
+    names(x)[names(x) == col_name] <- new_col_name
+  }
+  return(x)
+}
+
+trim_tbl_to_task_ids <- function(tbl, task_id_cols, hub_con,
+                                 call = rlang::caller_env()) {
+  if (is.null(task_id_cols) && is.null(hub_con)) {
+    cli::cli_abort(
+      c(
+        "x" = "Cannot trim to task IDs when both
+                  {.arg task_id_cols} and {.arg hub_con} are NULL",
+        "!" = "Please supply appropriate input to either
+                {.arg task_id_cols} or {.arg hub_con}."
+      ),
+      call = call
+    )
+  } else {
+    if (!is.null(task_id_cols)) {
+      if (!all(task_id_cols %in% names(tbl))) {
+        cli::cli_alert_warning(
+          "{.arg task_id_cols} value{?s} {.val
+                  {task_id_cols[!task_id_cols %in% names(tbl)]}
+                  } not a valid {.arg tbl} column name. Ignored."
+        )
+      }
+      task_id_cols <- task_id_cols[task_id_cols %in% names(tbl)]
+    } else {
+      if (!inherits(hub_con, "hub_connection")) {
+        cli::cli_abort("{.arg hub_con} must be a valid object of
+                                   class {.cls hub_connection}",
+          call = call
+        )
+      }
+      task_id_cols <- get_task_id_names(attr(hub_con, "config_tasks"))
```

Comment:
Is there a reason we don't need to do a similar check here to ensure that all `task_id_cols` named by the hub are in the `tbl` and handle the situation if not? it would seem plausible that there might be more task id columns defined by a hub than end up getting loaded, e.g. if task id columns vary by round.

Reply:
> it would seem plausible that there might be more task id columns defined by a hub than end up getting loaded, e.g. if task id columns vary by round.

Indeed because of this situation I have not implemented any checks. i.e. if a column in the data is a valid task_id across all rounds (i.e. from the full list of task_ids in the hub) I include it, if not I drop it.

Trying to validate round specific task_ids would be a much more complex operation that I would not recommend here.

Having said that I feel the data coming out of the hub should have columns for all available task ids (at least in the data). 

We could indeed do a check here if you like. We have three options for handling any task_ids that are found in the config but not in `tbl`:
- Ignore them but throw a warning that they are missing.
- add new  `NA` columns of the missing task ids.
- throw an error.

Let me know if you want this check implemented and which handling option you prefer.

Reply:
Also thanks for picking up on the `as_model_out_df` file name! I changed the name of the function half way through to better reflect the fact that we are using `tibbles` and missed changing the file name! Corrected in 8ea0de64120af64496708aa3a6ae2746f3d3b0cd

Reply:
The original reason I was asking was that I was thinking this could result in errors when we try to access possibly-non-existent columns at the end of the function:

```
> tmp <- tibble(a = 1:5, b = 2:6)
> tmp
# A tibble: 5 × 2
      a     b
  <int> <int>
1     1     2
2     2     3
3     3     4
4     4     5
5     5     6
> tmp[,letters[1:5]]
Error in `tmp[, letters[1:5]]`:
! Can't subset columns that don't exist.
✖ Columns `c`, `d`, and `e` don't exist.
Run `rlang::last_trace()` to see where the error occurred.
```

r.e. handling any task_ids that are found in the config but not in tbl: I think my preference would be to take a light touch here and do nothing:

- I think we should certainly validate these things when data are making their way into a hub by validating that the correct model task ids are present, on a per-round basis
- If I'm working with data for a hub that has two rounds (say, round 1 and round 2) that have different task id variables, it seems like it should be possible to work with model outputs for round 1 only without any need to raise warnings or introduce extra columns that only show up in round 2.

Reply:
re-reading your comment above that "The only situation where a task id defined in the config would not be present in data from a hub connection would be when no file in the hub has actually submitted data for that task id (i.e maybe all values are optional)" -- maybe I'm not understanding something about loading hub data.  If I only load data from one round, won't I only have columns for the task ids that are relevant for that round?

Reply:
@elray re: your question above, no you would have columns of `NA`s for any task IDs that are not present in the round but available within the general hub config. This is because the schema (and therefore columns) is set at the dataset level when opening an arrow connection.  You could use `remove_empty = TRUE` if you wanted to make sure any such columns are removed though.

Having read your first response however, the check you initially suggested is definitely something we should include as there are situations this might fail as is (complete oversight on my part).

Reply:
Implemented in a8436cbe278cb1b118dc1cdec420619ebdf46674 . Now check is performed and message issued if any trimming task id ignored regardless of which trimming method used.

Reply:
a8436cbe278cb1b118dc1cdec420619ebdf46674

Reply:
a8436cbe278cb1b118dc1cdec420619ebdf46674